proposals/0000-support-pun-free-code.md

```diff
@@ -0,0 +1,510 @@
+newtype ReaderT r m a = ReaderT {runReaderT :: r -> m a}
+newtype ExceptT e m a = ExceptT (m (Either e a))
+```
+
+However, as Haskell's type system evolves, the distinction between types and
+terms becomes blurry. For example, the `-XDataKinds` extension introduces the
+`'` syntax to select entities from the data namespace in a type-level context:
+
+```haskell
+r :: Rec Const '[ 'T ]   -- vinyl records
+r = Const :& RNil
+```
+
+Note that we had to qualify both the list syntax and the `T` data constructor
+with a tick.
+
+#### Syntactic Unification Principle (from #378)
+
+In the absence of punning, there is no difference between type-syntax and
+term-syntax.
+
+Syntactic Unification Principle implies that if the user is not using punning
+they can simply forget about the distinction between terms and types.
+
+#### Lexical Scoping Principle (from #378)
+
+For every occurrence of an identifier, it is possible to uniquely identify its
+binding site, without involving the type system.
+
+### Problem Statement
+
+As we step towards Dependent Haskell (with #378 acceptance), the distinction
+between types and terms becomes blurrier and blurrier and the need arises to
+use terms and types interchangeably. Indeed, we can begin to see this need in
+#281, which introduces "visible `forall`" syntax that lets us write functions
+like this:
+
+```haskell
+sizeOf :: forall a -> Sized a => Int
+
+sizeOfInt = sizeOf Int
+```
+
+But because of punning, this can result in ambiguity:
+
+```haskell
+data T = T
+
+sizeOfT = sizeOf T -- is 'T' the data type or the data constructor?
+```
+
+We can't use the type of `sizeOf` to determine which `T` to use because of
+*Lexical Scoping Principle*.
+
+#281 tackles this issue by defaulting `T` to a data constructor in this case
+(to keep compatibility with existing code) and introduces `type` syntactic marker.
+
+However, thanks to *Syntactic Unification Principle* (adhered by #281), if
+the user chooses to not use punning, there is no need to use this syntactic
+marker, resulting in less context-dependent and syntax cluttered code.
+
+Unfortunately, GHC has no good support for such style of programming:
+
+* There is no way for the user to know if some code uses punning without
+  manually reviewing it, in other words, usage of punning is silent.
+
+* There is no way to avoid punning if the user imports the code that makes the
+  use of it.
+  Consider this import statement: `import Data.Proxy`. It is impossible to
+  distinguish between `Proxy` the type and `Proxy` the data constructor.
+
+#### `'` syntax clash in `-XDataKinds` and `-XTemplateHaskell`
+
+`'` prefix syntax is used both by `-XDataKinds`, to promote a constructor,
+and in `-XTemplateHaskell`, to refer to a term-level identifier. For
+type-level identifiers `-XTemplateHaskell` uses `''`.
+
+This not only results in ambiguity when both of the extensions are enabled, it violates
+the *Syntactic Unification Principle* as users have to use `'` in
```

Comment:
I don't see this. That is, I don't think that `''` violates the SUP. The fact that `''` cannot be used in types violates the SUP: it's a difference between term-syntax and type-syntax. But I don't think that's the point you're making (because e.g. `case` violates the SUP in the same way). I think instead `''` violates the

**Namespace Ignorance Principle**. In the absence of punning, users never have to know whether an identifier is in the type namespace or the data namespace.

Now, `''` violates the NIP because we use `''` for type identifiers but `'` for data identifiers.